fix: Dnsrule bug for virtual Network link name

ChrisSidebotham opened the pull request:
## Description

update to virualtNetworkLinkName 

Closes #1463 

## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
|      [![avm.res.network.dns-forwarding-ruleset](https://github.com/ChrisSidebotham/bicep-registry-modules/actions/workflows/avm.res.network.dns-forwarding-ruleset.yml/badge.svg?branch=dnsrule-bug-1463)](https://github.com/ChrisSidebotham/bicep-registry-modules/actions/workflows/avm.res.network.dns-forwarding-ruleset.yml) |

## Type of Change

<!-- Use the check-boxes [x] on the options that are relevant. -->

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [x] Azure Verified Module updates:
  - [x] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [x] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

<!--  Please keep up to day with the contribution guide at https://aka.ms/avm/contribute/bicep -->

coolhome commented:
While this would solve the problem, I'm not sure if this is the route I would have taken.

If you compare against the `private-dns-zone` there is also a concept of "Virtual Network Links". These have convention defaults (e.g. `<vnet name>-vnetlink`) and the ability to override. I think this is more robust to allow users to take convention but also override if necessary.  Just my two cents.

https://github.com/Azure/bicep-registry-modules/blob/05ac4c19fc8a45a0f54ef76975ed25576607cfc5/avm/res/network/private-dns-zone/main.bicep#L205-L217

ChrisSidebotham commented:
> While this would solve the problem, I'm not sure if this is the route I would have taken.
> 
> If you compare against the `private-dns-zone` there is also a concept of "Virtual Network Links". These have convention defaults (e.g. `<vnet name>-vnetlink`) and the ability to override. I think this is more robust to allow users to take convention but also override if necessary. Just my two cents.
> 
> https://github.com/Azure/bicep-registry-modules/blob/05ac4c19fc8a45a0f54ef76975ed25576607cfc5/avm/res/network/private-dns-zone/main.bicep#L205-L217

@cool-home am I missed something here? 

In the current iteration you can override the value of the link name by supplying `forwardingRule.name`. Howver I am more than happy to allign this with the privateDnsZone Implementation of `vnetLinks` and will get this updated.

ChrisSidebotham commented:
> ```bicep
> virtualNetworkResourceId
> ```

@coolhome - Forgive me for the confusion here. I have just updated this now. I have also added a UDT for the virtualNetworkLinks array so it has the added intelisense to support this also.

coolhome commented on avm/res/network/dns-forwarding-ruleset/virtual-network-link/main.bicep:
```diff
@@ -19,7 +19,7 @@ resource dnsForwardingRuleset 'Microsoft.Network/dnsForwardingRulesets@2022-07-0
 
 resource virtualNetworkLink 'Microsoft.Network/dnsForwardingRulesets/virtualNetworkLinks@2022-07-01' = {
-  name: name ?? '${last(split(virtualNetworkResourceId, '/'))}-vnetlink'
+  name: name ?? '${substring(uniqueString(subscription().id),0,8)}-${substring(uniqueString(resourceGroup().id),0,8)}-${last(split(virtualNetworkResourceId, '/'))}-vnetlink'
```
Maybe this should be reverted since the user can accept original behavior or override using new UDT.

ChrisSidebotham replied:
Reverted 💪

coolhome reviewed (approved):
This solves my problem and I think enhances the module :) Thank you!

ChrisSidebotham commented:
Pipeline failure is for timing issue with resource removal. @AlexanderSehr could you approve this?

ChrisSidebotham commented:
Pipeline failiure is for timing issue with resource removal. @AlexanderSehr could you approve this?

AlexanderSehr commented on avm/res/network/dns-forwarding-ruleset/main.bicep:
```diff
@@ -25,7 +25,7 @@ param dnsForwardingRulesetOutboundEndpointResourceIds array
 
 @description('Optional. Array of virtual network links.')
-param vNetLinks array?
+param virtualNetworkLinks virtualNetworkLinkType?
```
```suggestion
param virtualNetworkLinks virtualNetworkLinkType
```

Not both the type and parameter must be nullable. One is enough 😉

AlexanderSehr replied:
Once that's done and the `Set-AVMModule` script run (you heard that right, it's a rap), I'm happy to approve

ChrisSidebotham replied:
Done please rereview @AlexanderSehr
